src/core/service.c

```diff
@@ -1250,23 +1271,98 @@ static int service_coldplug(Unit *u) {
         return 0;
 }
 
+static int connect_unix_probe(Service *s, const OpenFile *of, int ofd) {
+        union sockaddr_union addr = {
+                .un.sun_family = AF_UNIX,
+        };
+        socklen_t sa_len;
+        int socket_types[] = { SOCK_DGRAM, SOCK_STREAM, SOCK_SEQPACKET };
+        int r;
+
+        assert(s);
+        assert(of);
+        assert(ofd >= 0);
+
+        r = sockaddr_un_set_path(&addr.un, FORMAT_PROC_FD_PATH(ofd));
+        if (r < 0)
+                return log_unit_error_errno(UNIT(s), r, "Failed to set sockaddr for %s: %m", of->path);
+
+        sa_len = r;
+
+        for (size_t i = 0; i < ELEMENTSOF(socket_types); i++) {
+                _cleanup_close_ int fd = -1;
+
+                fd = socket(AF_UNIX, socket_types[i] | SOCK_CLOEXEC, 0);
+                if (fd < 0)
+                        return log_unit_error_errno(
+                                        UNIT(s), errno, "Failed to create socket for %s: %m", of->path);
+
+                r = RET_NERRNO(connect(fd, &addr.sa, sa_len));
+                if (r >= 0)
+                        return TAKE_FD(fd);
+
+                if (r != -EPROTOTYPE)
+                        return log_unit_error_errno(UNIT(s), r, "Failed to create socket for %s: %m", of->path);
+        }
+
+        return log_unit_error_errno(UNIT(s), SYNTHETIC_ERRNO(EINVAL), "Failed to connect to socket sun_path %s sun_family %d: %m", addr.un.sun_path, addr.un.sun_family);
+}
+
+static int get_open_file_fd(Service *s, const OpenFile *of) {
+        struct stat st;
+        _cleanup_close_ int fd = -1, ofd = -1;
+        int flags = O_RDWR;
+
+        assert(s);
+        assert(of);
+
+        ofd = open(of->path, O_PATH | O_CLOEXEC);
+        if (fstat(ofd, &st) < 0)
+                return log_error_errno(errno, "Failed to stat %s: %m", of->path);
+
+        if (S_ISSOCK(st.st_mode)) {
+                fd = connect_unix_probe(s, of, ofd);
+                if (fd < 0)
+                        return fd;
+
```

Comment:
maybe, if `FLAGS_SET(of->flags, OPENFILE_RDONLY)` holds issue `shutdown(fd, SHR_WR)` on the socket?